### PR TITLE
Reduce docker size

### PR DESCRIPTION
**Pull request by SkySkimmer**
Size changes in the commit messages are gotten using https://github.com/wagoodman/dive
I think docker applies some compression so expect to see smaller changes in https://gitlab.inria.fr/coq/coq/container_registry/2437

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Additional info:
/usr/lib/x86_64-linux-gnu is pretty big, notably the libllvm.so is at around 100MB. It seems to be brought in by the gtk packages (ie just for coqide).

latex seems to add about 300MB of stuff.

I may have miscounted changes somewhere as I feel like the numbers don't add up x_x

**Comment by SkySkimmer** [comment]
<table>
<tr>
 <td>
 <td>previous (2023-11-30)
 <td>this PR
<tr>
 <td>"old"
 <td>1.05 GiB
 <td>773.81 MiB (-28%)
<tr>
 <td>"edge"
 <td>1.32 GiB
 <td>1.01 GiB (-23%)
</table>
Let's see how light CI does, then if it works we will run full CI.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
quickchick error looks unrelated to this PR although I'm not sure what's going on there
https://github.com/QuickChick/QuickChick/pull/348 removed the file mentioned in the error but its CI is green
~~~
*** Error: plugin/tactic_quickchick.mli: No such file or directory
~~~
cc @lemonidas, @Lysxia, @liyishuai

File: dev/ci/docker/edge_ubuntu/Dockerfile [hunk 1]
```diff
@@ -15,16 +15,15 @@ RUN apt-get update -qq && apt-get install --no-install-recommends -y -qq \
...
         # Dependencies of source-doc and coq-makefile
         texlive-latex-extra texlive-science tipa \
         # Dependencies of HB (test suite)
         wdiff \
         # Required to get the wget step to succeed
         ca-certificates \
         # Required for fiat-crypto and Coqtail
-        python-is-python3
+        python-is-python3 \
+      && rm -rf /var/lib/apt /usr/share/doc
```
Comment (ejgallego) [inline dev/ci/docker/edge_ubuntu/Dockerfile:26]: Removing /var/lib/apt may prove too heavy handed, if for example one want to log into the image to debug apt installs; for 60MiB I would keep it, but do as you wish.

Note that several subdirs in `var/lib/apt` can be indeed removed without problem (such as `/var/lib/apt/lists`), which is usually the culprit for size problem there.

**Comment by silene** [comment]
> Especially the gappa removal probably isn't actually 600MB (but is still hundreds of MB)

That would be surprising. The whole Docker image used by Gappa's own CI is only 165MB, and that is despite including the whole Python3/Sphinx toolchain for generating the documentation.

**Comment by SkySkimmer** [comment]
I see https://gitlab.inria.fr/gappa/coq/-/blob/master/misc/Dockerfile installs libboost-dev not libboost-all-dev, maybe that's the difference.

**Review by ejgallego — APPROVED** [review]
LGTM, tho maybe this is a bit too radical in a couple of places, I'd suggest:
- don't delete `/var/lib/apt`, but /var/lib/apt/lists` which is the large directory
- maybe keep the .cmt files? How much we are saving by this anyways? If that's not too much, I think the tradeoff of deviating our images from standard practice in the OCaml world is way too risky, as tools may expect the cmt files to be there.

**Comment by ejgallego** [comment]
In my system the overhead of cmt files is not trivial indeed.

**Comment by ppedrot** [comment]
@SkySkimmer I think it doesn't hurt to merge the commits from this PR incrementally. Otherwise we're going to nitpick the changes, and in the meantime we have gigabytes of wasted throughput on our CI.

**Comment by JasonGross** [comment]
Does it make sense to base the docker images on Alpine Linux, which has a base image size of only 5 MB

**Comment by SkySkimmer** [comment]
The ubuntu base image is around 30MB, the vast majority of the size is the packages we install.

**Comment by SkySkimmer** [comment]
Sizes after rebase are pretty much the same as before rebase

**Comment by ejgallego** [comment]
@SkySkimmer should we undraft (and merge) this?

If the `.cmt` change creates problems we can always revert it later.

**Comment by SkySkimmer** [comment]
I forgot this was draft
@coqbot run full ci

**Comment by ejgallego** [comment]
@coqbot: merge now

**Pull request merged**